spec/lib/miq_preloader_spec.rb

```diff
@@ -4,29 +4,58 @@
       ems = FactoryBot.create(:ems_infra)
       expect(ems.vms).not_to be_loaded
       expect { preload(ems, :vms) }.to make_database_queries(:count => 1)
+
       expect(ems.vms).to be_loaded
       expect { preload(ems, :vms) }.to_not make_database_queries
+      expect { ems.vms.size }.not_to make_database_queries
     end
 
-    it "preloads from an array" do
+    it "preloads an array" do
       emses = FactoryBot.create_list(:ems_infra, 2)
       expect { preload(emses, :vms) }.to make_database_queries(:count => 1)
-      expect(emses[0].vms).to be_loaded
+
+      expect(emses.first.vms).to be_loaded
+      expect { emses.first.vms.size }.not_to make_database_queries
     end
 
-    it "preloads from an association" do
+    it "preloads a relation (record is a relation)" do
+      FactoryBot.create_list(:vm, 2, :ext_management_system => FactoryBot.create(:ems_infra))
+      emses = ExtManagementSystem.all
+      expect { preload(emses, :vms) }.to make_database_queries(:count => 2)
+
+      expect { expect(emses.first.vms.size).to eq(2) }.not_to make_database_queries
+    end
+
+    it "preloads with a relation (records is a relation)" do
       ems = FactoryBot.create(:ems_infra)
       FactoryBot.create_list(:vm, 2, :ext_management_system => ems)
 
-      emses = ExtManagementSystem.all
-      expect { preload(emses, :vms) }.to make_database_queries(:count => 2)
+      emses = ExtManagementSystem.all.load
+      vmses = Vm.where(:ems_id => emses.select(:id))
+      expect { preload(emses, :vms, vmses) }.to make_database_queries(:count => 1)
     end
 
     def preload(*args)
       MiqPreloader.preload(*args)
     end
   end
 
+  describe ".preload_reverse" do
+    it "loads belongs_to in a reverse" do
+      ems = FactoryBot.create(:ems_infra)
+      vms = FactoryBot.create_list(:vm, 2, :ext_management_system => ems)
+
+      vms.each(&:reload)
+      expect(vms.first.association(:ext_management_system)).not_to be_loaded
+      expect { vms.map(&:ext_management_system) }.to make_database_queries
+
+      vms.each(&:reload)
+      expect { MiqPreloader.preload_reverse(ems, :vms, vms) }.not_to make_database_queries
```

Comment:
Oh but now I'm seeing :ems in the test - this is confusing.  I think I know an interface that would make this less complicated, but I need to discuss.